bash.sh

```diff
@@ -12,40 +12,83 @@
 #
 # The only proper way to do this is by looping through each item. Other
 # solutions claim to do this using string matching, but they are often
-# dangerous because don't handle edge cases.
+# dangerous because they don't handle edge cases, e.g. may be susceptible to
+# partial match or may not properly handle elements containing newlines.
 #
 
 #
 # Returns 0 if item is in the array; 1 otherwise.
 #
-# $1: The array value to search for
-# $@: The array values, e.g. "${myarray[@]}"
+# $1: name of the array variable to search
+# $2: array item to search for
 #
 array_contains() {
-    local item=$1; shift
-    for val in "$@"; do
+    local arr_name=$1
+    local item=$2
+
+    eval "local -a _tmp_arr=(\"\${${arr_name}[@]}\")"
+
+    for val in "${_tmp_arr[@]}"; do
         if [ "$val" == "$item" ]; then
             return 0
         fi
     done
+
+    return 1
+}
+
+
+#
+# Returns 0 if regexp matches any item in the array; 1 otherwise.
+#
+# $1: name of the array variable to search
+# $2: regexp to match
+#
+array_contains_regexp() {
+    local arr_name=$1
+    local item=$2
+
+    eval "local -a _tmp_arr=(\"\${${arr_name}[@]}\")"
+
+    for val in "${_tmp_arr[@]}"; do
+        if [[ "$val" =~ $item ]]; then
+            return 0
+        fi
+    done
+
     return 1
 }
 
 # ### Array filter
 
 #
-# Return all elements of an array with the specified item removed.
+# Remove elements from the named array. Iterate over each element and call
+# named function. If the function returns false, the element is removed.
 #
-# $1: The array value to remove
-# $@: The array values, e.g. "${myarray[@]}"
+# $1: name of the array variable
+# $@: name of the function to call plus any arguments; element to test will be
```

Comment:
why not call out func_name as a separate argument? 

That's what the code does too.